Source/Provers/SMTLib/SMTLibBatchTheoremProver.cs

```diff
@@ -0,0 +1,318 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Diagnostics;
+using Microsoft.Boogie.VCExprAST;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Boogie.SMTLib
+{
+  public class SMTLibBatchTheoremProver : SMTLibProcessTheoremProver
+  {
+    private SMTLibSolver Process;
+    private bool CheckSatSent;
+    private int resourceCount;
+    private Model errorModel;
+
+    [NotDelayed]
+    public SMTLibBatchTheoremProver(SMTLibOptions libOptions, ProverOptions options, VCExpressionGenerator gen,
+      SMTLibProverContext ctx) : base(libOptions, options, gen, ctx)
+    {
+      if (usingUnsatCore) {
+        throw new NotSupportedException("Batch mode solver interface does not support unsat cores.");
+      }
+    }
+
+    public override Task GoBackToIdle()
+    {
+      return Task.CompletedTask;
+    }
+
+    private void SetupProcess()
+    {
+      Process?.Close();
+      Process = options.Solver == SolverKind.NoOpWithZ3Options ? new NoopSolver() : new SMTLibProcess(libOptions, options);
+      Process.ErrorHandler += HandleProverError;
+      CheckSatSent = false;
+    }
+
+    public override int FlushAxiomsToTheoremProver()
+    {
+      // we feed the axioms when BeginCheck is called.
+      return 0;
+    }
+
+    public override void Close()
+    {
+      base.Close();
+      Process?.Close();
+      Process = null;
+    }
+
+    public override void BeginCheck(string descriptiveName, VCExpr vc, ErrorHandler handler)
+    {
+      SetupProcess();
+      FullReset(gen);
+
+      if (options.LogFilename != null && currentLogFile == null) {
+        currentLogFile = OpenOutputFile(descriptiveName);
+        currentLogFile.Write(common.ToString());
+      }
+
+      PrepareCommon();
+      FlushAxioms();
+
+      OptimizationRequests.Clear();
+
+      string vcString = "(assert (not\n" + VCExpr2String(vc, 1) + "\n))";
+      FlushAxioms();
+
+      Push();
+      SendVCAndOptions(descriptiveName, vcString);
+      SendOptimizationRequests();
+
+      FlushLogFile();
+
+      Process.NewProblem(descriptiveName);
+
+      SendCheckSat();
+      Pop();
+
+      FlushLogFile();
+    }
+
+    public override void Reset(VCExpressionGenerator generator)
+    {
+    }
+
+    public override void FullReset(VCExpressionGenerator generator)
+    {
+      if (options.Solver == SolverKind.Z3 || options.Solver == SolverKind.NoOpWithZ3Options)
```

Comment:
Odd that this is conditional on Z3

Reply:
That's inherited from the original code, so I'm inclined to leave it that way.